### PR TITLE
HYDRA-691 : Move mayaUsd code for managing stages inside mayaHydra

### PR DESCRIPTION
**Pull request by lanierd-adsk**

### CODE REVIEW

**Comment by lanierd-adsk** [comment]
FYI : @roopavr-adsk @lilike-adsk @debloip-adsk

**Comment by lanierd-adsk** [comment]
This won't be merged before our release is done.

**Comment by lanierd-adsk** [comment]
FYI : @lilike-adsk , @roopavr-adsk , @debloip-adsk

File: doc/build.md [hunk 1]
```diff
@@ -80,6 +80,14 @@ There are four arguments that must be passed to the script:
 |  --devkit-location    | directory where Maya devkit is installed.                                             |
 | workspace_location    | directory where the project use as a workspace to build and install plugin/libraries  |
 
+Optional arguments are :
+| Flags                 | Description                                                                                         |
+|--------------------   |---------------------------------------------------------------------------------------------------- |
+|   --mayausd-location  | directory where the mayaUSD.mod file is. By providing this, you will enable more features for       |
+|                       | usd stages when using an hydra render delegate, such as : hide/delete the stage when the proxy shape|
+|                       | node is hidden/deleted, or applying a transform on the proxy shape node will apply it on the stage. |
```
Comment (ppt-adsk) [inline doc/build.md]: This sounds awkward, as if the version without --mayausd-location is buggy.  Would such a configuration be useful in production?  It would seem far preferable for customers to have a fully-working integration, or nothing at all.  Otherwise, QA has to test 2 configurations for USD, one which is arguably broken.

File: lib/mayaHydra/hydraExtensions/CMakeLists.txt [hunk 2]
```diff
@@ -30,31 +30,54 @@ set(HEADERS
...
-        MAYAHYDRALIB_EXPORT
-        $<$<BOOL:${IS_MACOSX}>:OSMac_>
-        # Copy-pasted from lib/mayaUsd/CMakeLists.txt
-)
-
+if(DEFINED MAYAUSD_LOCATION)
+    target_compile_definitions(${TARGET_NAME} 
+        PRIVATE 
+            MAYAHYDRALIB_EXPORT
+            MAYAHYDRALIB_MAYAUSD_IS_USED_TO_BUILD
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/CMakeLists.txt]: A bit wordy for a boolean variable.  How about:
```suggestion
            MAYAHYDRALIB_MAYAUSD_ENABLED
```

File: lib/mayaHydra/hydraExtensions/CMakeLists.txt [hunk 3]
```diff
@@ -30,31 +30,54 @@ set(HEADERS
...
 # compiler configuration
 # -----------------------------------------------------------------------------
-target_compile_definitions(${TARGET_NAME} 
-    PRIVATE 
-        MAYAHYDRALIB_EXPORT
-        $<$<BOOL:${IS_MACOSX}>:OSMac_>
-        # Copy-pasted from lib/mayaUsd/CMakeLists.txt
-)
-
+if(DEFINED MAYAUSD_LOCATION)
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/CMakeLists.txt]: I don't think you need this conditional, could simply be
```suggestion
$<$<BOOL:${MAYAUSD_LOCATION}>:MAYAHYDRALIB_MAYAUSD_ENABLED>
```

File: lib/mayaHydra/hydraExtensions/CMakeLists.txt [hunk 4]
```diff
@@ -30,31 +30,54 @@ set(HEADERS
...
-            ${PXR_INCLUDE_DIRS}
-            ${BOOST_INCLUDE_DIR}
-            ${CMAKE_BINARY_DIR}/include
-        PRIVATE
-            $<$<BOOL:${UFE_FOUND}>:${UFE_INCLUDE_DIR}>
-    )
+# -----------------------------------------------------------------------------
+# include directories
+# -----------------------------------------------------------------------------
+if(DEFINED MAYAUSD_LOCATION)
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/CMakeLists.txt]: See above using a CMake generator expression.

File: lib/mayaHydra/hydraExtensions/CMakeLists.txt [hunk 5]
```diff
@@ -94,14 +116,43 @@ endif()
     # -----------------------------------------------------------------------------
     # link libraries
     # -----------------------------------------------------------------------------
-if(BUILD_MAYAUSD_LIBRARY)
+if(DEFINED MAYAUSD_LOCATION)
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/CMakeLists.txt]: Again, a generator expression should be used.

File: lib/mayaHydra/hydraExtensions/sceneIndex/CMakeLists.txt [hunk 6]
```diff
@@ -32,6 +33,7 @@ set(HEADERS
     mayaHydraSceneIndexDataFactoriesSetup.h
     mayaHydraMayaFilteringSceneIndexData.h
     mayaHydraMayaFilteringSceneIndexDataConcreteFactory.h
+    mayaHydraMayaUsdProxyShapeSceneIndex.h
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/CMakeLists.txt]: With our recent changes to the coding guidelines I believe these file names should be:
```suggestion
    mhMayaUsdProxyShapeSceneIndex.h
```

File: lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraMayaUsdProxyShapeSceneIndex.h [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+//Usd/Hydra headers
+#include <pxr/base/tf/declarePtrs.h>
+#include <pxr/usdImaging/usdImaging/stageSceneIndex.h>
+#include <pxr/usdImaging/usdImaging/sceneIndices.h>// In USD 23.11+
+#include <pxr/imaging/hd/filteringSceneIndex.h>
+
+//Maya headers
+#include <maya/MObjectHandle.h>
+#include <maya/MMEssage.h>
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraMayaUsdProxyShapeSceneIndex.h]: This won't compile on Linux, where the file system is case-sensitive.
```suggestion
#include <maya/MMessage.h>
```

File: lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraMayaUsdProxyShapeSceneIndex.h [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+namespace MAYAHYDRA_NS_DEF {
+
+class MayaUsdProxyShapeSceneIndex;
+TF_DECLARE_WEAK_AND_REF_PTRS(MayaUsdProxyShapeSceneIndex);
+
+/// <summary>
+/// Simply wraps single stage scene index for initial stage assignment and population
+/// </summary>
+class MayaUsdProxyShapeSceneIndex : public HdSingleInputFilteringSceneIndexBase
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraMayaUsdProxyShapeSceneIndex.h]:
```suggestion
class MhMayaUsdProxyShapeSceneIndex : public HdSingleInputFilteringSceneIndexBase
```

File: build.py [hunk 9]
```diff
@@ -556,7 +556,7 @@ def Package(context):
                     help="Directory where Maya is installed.")
 
 parser.add_argument("--mayausd-location", type=str,
-                    help="Directory where MayaUsd is installed.")
+                    help="Directory where mayaUSD.mod file is.")
```
Comment (seando-adsk) [inline build.py]: Actually I think the previous help was correct. This flag should be used to specify where the MayaUsd install is. Then you should create a `cmake/modules/FindMayaUsd.cmake` file where search for the new api library you want to link with and the include folder (setting two new cmake variables for these (similar to FindUfe.cmake), such as: MAYAUSD_INCLUDE_DIR and MAYAUSD_API_LIBRARY.

I think you could also add cmake code in this new FindMayaUsd.cmake that looks for the mayaUsd.mod file and if found adds the path to MAYA_MODULE_PATH.

File: doc/build.md (hunk 1 above)
Comment (seando-adsk) [inline doc/build.md]: I don't think its where the .mod is, but instead simply where MayaUsd is (the root folder).

File: lib/mayaHydra/mayaPlugin/renderOverride.cpp [hunk 10]
```diff
@@ -60,10 +59,6 @@
 
 #include <ufeExtensions/Global.h>
 
-#if defined(MAYAUSD_VERSION)
-#include <mayaUsd/render/px_vp20/utils.h>
-#include <mayaUsd/utils/hash.h>
-#else
 namespace MayaUsd {
 // hash combiner taken from:
 // http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0814r0.pdf
```
Comment (ppt-adsk) [inline lib/mayaHydra/mayaPlugin/renderOverride.cpp]: This is all dead code, and should be removed.  We should never have code in the MayaUsd namespace in mayaHydra.

File: lib/mayaHydra/hydraExtensions/CMakeLists.txt [hunk 11]
```diff
@@ -30,31 +30,54 @@ set(HEADERS
...
+            # Copy-pasted from lib/mayaUsd/CMakeLists.txt
+    )
+else()
+    target_compile_definitions(${TARGET_NAME} 
+        PRIVATE 
+            MAYAHYDRALIB_EXPORT
+            $<$<BOOL:${IS_MACOSX}>:OSMac_>
+            # Copy-pasted from lib/mayaUsd/CMakeLists.txt
+    )
+endif()
```
Comment (seando-adsk) [inline lib/mayaHydra/hydraExtensions/CMakeLists.txt]: Once you create the FindMayaUsd.cmake it will set MAYAUSD_FOUND (to TRUE for cmake). Then this can be simplified with a cmake generator expression.
```suggestion
target_compile_definitions(${TARGET_NAME} 
    PRIVATE 
        MAYAHYDRALIB_EXPORT
        $<$<BOOL:${IS_MACOSX}>:OSMac_>
        $<$<BOOL:${MAYAUSD_FOUND}>:MAYAHYDRALIB_MAYAUSD_IS_USED_TO_BUILD>
        # Copy-pasted from lib/mayaUsd/CMakeLists.txt
)
```

File: lib/mayaHydra/hydraExtensions/CMakeLists.txt [hunk 12]
```diff
@@ -30,31 +30,54 @@ set(HEADERS
...
+            PUBLIC
+                ${MAYA_INCLUDE_DIRS}
+                ${PYTHON_INCLUDE_DIR}
+                ${PXR_INCLUDE_DIRS}
+                ${BOOST_INCLUDE_DIR}
+                ${CMAKE_BINARY_DIR}/include
+            PRIVATE
+                $<$<BOOL:${UFE_FOUND}>:${UFE_INCLUDE_DIR}>
+        )
+     endif()
```
Comment (seando-adsk) [inline lib/mayaHydra/hydraExtensions/CMakeLists.txt]: Can simplify with generator expression.
```suggestion
    target_include_directories(${TARGET_NAME} 
        PUBLIC
            ${MAYA_INCLUDE_DIRS}
            ${PYTHON_INCLUDE_DIR}
            ${PXR_INCLUDE_DIRS}
            ${BOOST_INCLUDE_DIR}
            ${CMAKE_BINARY_DIR}/include
        PRIVATE
            $<$<BOOL:${UFE_FOUND}>:${UFE_INCLUDE_DIR}>
            $<$<BOOL:${MAYAUSD_FOUND}>:${MAYAUSD_INCLUDE_DIR}>
    )
```

File: lib/mayaHydra/mayaPlugin/renderOverride.cpp [hunk 13]
```diff
@@ -1099,12 +1097,24 @@ void MtohRenderOverride::_PopulateSelectionList(
...
+                    if (foundPlace == std::string::npos) {
+                        TF_CODING_ERROR("pickedPathAsString.find(sceneIndexPathPrefixAsString) returned std::string::npos !");
+                        return;
+                    }
+
+                    //Keep only the right part of the string after the prefix
+                    const size_t placeToKeep = foundPlace + sceneIndexPathPrefixAsString.length();
+                    localPathAsString = pickedPathAsString.substr(placeToKeep, pickedPathAsString.length() - placeToKeep);
+
+                    const SdfPath localPath(localPathAsString);
```
Comment (ppt-adsk) [inline lib/mayaHydra/mayaPlugin/renderOverride.cpp]: I can see that you went to great lengths to replace the existing code, but on reading through I don't understand why we need convert to string / find prefix / replace prefix.  This seems to be way too much work.  Additionally, the SdfPath documentation recommends against converting to strings to do processing:
https://openusd.org/release/api/class_sdf_path.html#details

File: lib/mayaHydra/hydraExtensions/CMakeLists.txt [hunk 14]
```diff
@@ -94,14 +116,43 @@ endif()
     # -----------------------------------------------------------------------------
     # link libraries
     # -----------------------------------------------------------------------------
-if(BUILD_MAYAUSD_LIBRARY)
+if(DEFINED MAYAUSD_LOCATION)
+    find_library(MAYAUSD_API_LIBRARY NAMES mayaUsdAPI PATHS ${MAYAUSD_LOCATION}/lib NO_DEFAULT_PATH)
+    if(NOT MAYAUSD_API_LIBRARY)
+        message(WARNING "Could not find mayaUsdAPI library.")
+    endif()
```
Comment (seando-adsk) [inline lib/mayaHydra/hydraExtensions/CMakeLists.txt]: This should be in the new FindMayaUsd.cmake

Comment (seando-adsk) [reply lib/mayaHydra/hydraExtensions/CMakeLists.txt]: Can use generator expression again to add the library (if found). No need for the `if(DEFINED MAYAUSD_LOCATION)`

            $<$<BOOL:${MAYAUSD_FOUND}>:${MAYAUSD_API_LIBRARY}>

Comment (seando-adsk) [reply lib/mayaHydra/hydraExtensions/CMakeLists.txt]: I'm pretty sure the `if(DEFINED MAYAUSD_VERSION)` can be simplified down to a single generator expression line as well. This might/should work:
`$<$<BOOL:${MAYAUSD_FOUND}>:MAYAUSD_VERSION=${MAYAUSD_VERSION}>`

**Review by seando-adsk — CHANGES_REQUESTED** [review]

File: lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraMayaUsdProxyShapeSceneIndex.cpp [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "mayaHydraMayaUsdProxyShapeSceneIndex.h"
+
+#if defined(MAYAHYDRALIB_MAYAUSD_IS_USED_TO_BUILD)
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraMayaUsdProxyShapeSceneIndex.cpp]: I would be more assertive and do:
```suggestion
#if not defined(MAYAHYDRALIB_MAYAUSD_IS_USED_TO_BUILD)
#error MAYAHYDRALIB_MAYAUSD_ENABLED must be defined!
#endif
```

File: lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraMayaUsdProxyShapeSceneIndex.cpp [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+            _populated = true;
+        }
+    }
+}
+
+Ufe::Path MayaUsdProxyShapeSceneIndex::InterpretRprimPath(
+    const HdSceneIndexBaseRefPtr& sceneIndex,
+    const SdfPath&                path)
+{
+    if (MayaUsdProxyShapeSceneIndexRefPtr proxyShapeSceneIndex = TfStatic_cast<MayaUsdProxyShapeSceneIndexRefPtr>(sceneIndex)) {
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraMayaUsdProxyShapeSceneIndex.cpp]: This line is suspicious.  TfStatic_cast does not do a dynamic cast; it is therefore equivalent to doing a nullptr check on the input sceneIndex.  I wonder if a dynamic cast was intended?

File: lib/flowViewport/API/fvpFilteringSceneIndexClient.h [hunk 17]
```diff
@@ -58,7 +58,7 @@ namespace FVP_NS_DEF
         *  @param[in]      rendererNames is the names of the renderers you want this client to be associated to.
         *                  If there are several, separate them with for example a coma, like "GL, Arnold", we actually look for the renderer name in this string.
         *                  If you want your client to work on any renderer please use FvpViewportAPITokens->allRenderers.
-        *  @param[in]      dccNode is a MObject* for Maya, if you provide the pointer value, then we automatically track some events such as visibility changed, 
+        *  @param[in]      dccNode is a MObject* DAG node for Maya, if you provide the pointer value, then we automatically track some events such as visibility changed,
```
Comment (ppt-adsk) [inline lib/flowViewport/API/fvpFilteringSceneIndexClient.h:61]: If it's a Dag node, shouldn't we be passing in an MDagPath?

**Review by ppt-adsk — COMMENTED** [review]
We need a test for this.

File: lib/flowViewport/API/fvpFilteringSceneIndexClient.h (hunk 17 above)
Comment (ppt-adsk) [reply lib/flowViewport/API/fvpFilteringSceneIndexClient.h:61]: I'm concerned about our support for DAG nodes.  To correctly support Maya instances a DAG path must be used.  Furthermore, since visibility is hierarchical, and therefore can be changed by reparenting a node, tracking visibility changes needs to deal with a DAG path and track changes to it.

File: CMakeLists.txt [hunk 18]
```diff
@@ -127,6 +127,13 @@ else()
     message(STATUS "UFE not found.")
 endif()
 
+find_package(MAYAUSD OPTIONAL_COMPONENTS)
```
Comment (ppt-adsk) [inline CMakeLists.txt]: Nit-pick: why the all-caps naming?  UFE and USD packages are uppercase, because their name are uppercase, but MayaUsd is not --- e.g. look at the Python and Qt6 package names

File: test/lib/mayaUsd/render/mayaToHydra/testMayaUsdAPIUsage.py
Comment (ppt-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/testMayaUsdAPIUsage.py:1]: Thanks for the test!

**Review by ppt-adsk — DISMISSED** [review]
Only minor comments, approving.  From an architecture standpoint I am concerned about our use of MObject rather than MDagPath, but this is a wider concern, and should not block this pull request.

**Review by ppt-adsk — DISMISSED** [review]

File: CMakeLists.txt [hunk 19]
```diff
@@ -127,6 +127,13 @@ else()
     message(STATUS "UFE not found.")
 endif()
 
+find_package(MayaUsd OPTIONAL_COMPONENTS)
```
Comment (seando-adsk) [inline CMakeLists.txt]: I don't think you need the `OPTIONAL_COMPONENTS` - that is used when you want to specific a list of components which are optional. Since you didn't add `REQUIRED` this package is optional.

File: cmake/modules/FindMayaUsd.cmake [hunk 20]
```diff
@@ -0,0 +1,90 @@
+#
+# Simple module to find MayaUsd.
+#
+# This module searches for a valid MayaUsd installation.
+# It searches for MayaUsd's libraries and include header files.
+#
+# Variables that will be defined:
+# MAYAUSD_FOUND           Defined if a MayaUsd installation has been detected
```
Comment (seando-adsk) [inline cmake/modules/FindMayaUsd.cmake]:
```suggestion
# MayaUsd_FOUND           Defined if a MayaUsd installation has been detected
```

File: cmake/modules/FindMayaUsd.cmake [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+
+find_path(MAYAUSD_LIBRARY_DIR
+        mayaUsd.lib
+    HINTS
+        ${MAYAUSD_LOCATION}
+    PATH_SUFFIXES
+        lib/
+    DOC
+        "MayaUsd libraries path"
+)
```
Comment (seando-adsk) [inline cmake/modules/FindMayaUsd.cmake]: Why do you need to find this path? You just need the API one (below) - right? If you do need this one, then you'll have to adjust for linux/mac because the extension is not .lib on those platforms.

File: cmake/modules/FindMayaUsd.cmake [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+
+
+# Handle the QUIETLY and REQUIRED arguments and set MAYAUSD_FOUND to TRUE if
+# all listed variables are TRUE.
+include(FindPackageHandleStandardArgs)
+
+find_package_handle_standard_args(MAYAUSD
+    REQUIRED_VARS
+        MAYAUSD_INCLUDE_DIR
+        MAYAUSD_LIBRARY_DIR
```
Comment (seando-adsk) [inline cmake/modules/FindMayaUsd.cmake]: The required library is the API, so that should be the required variable, along with the include dir. These will only be set when MayaUsd_FOUND is true. That way you use `MayaUsd_FOUND` in the generator expression BOOL test.
```suggestion
        MAYAUSDAPI_LIBRARY
```

File: cmake/modules/FindMayaUsd.cmake [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+)
+
+if(MAYAUSD_FOUND)
+    message(STATUS "MayaUsd found : ${MAYAUSD_LOCATION}")
+    message(STATUS "MayaUsd include dir: ${MAYAUSD_INCLUDE_DIR}")
+    message(STATUS "MayaUsd libraries dir: ${MAYAUSD_LIBRARY_DIR}")
+    message(STATUS "MayaUsdAPI.lib fullpath : ${MAYAUSDAPI_LIBRARY}")
```
Comment (seando-adsk) [inline cmake/modules/FindMayaUsd.cmake]: I would not specify .lib - that is Windows specific.
```suggestion
    message(STATUS "MayaUsdAPI library : ${MAYAUSDAPI_LIBRARY}")
```

File: cmake/modules/FindMayaUsd.cmake [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+    #Add MAYAUSD_MOD_PATH (the path where maya USD .mod file is) to the MAYA_MODULE_PATH
+    
+    # Get the current value of the environment variable
+    set(CURRENT_MAYA_MODULE_PATH $ENV{MAYA_MODULE_PATH})
+
+    # Append the new path to the current value
+    set(MAYA_MODULE_PATH "${CURRENT_MAYA_MODULE_PATH}:${MAYAUSD_MOD_PATH}")
+
+    # Export the new value to the environment
+    set(ENV{MAYA_MODULE_PATH} ${MAYA_MODULE_PATH})
```
Comment (seando-adsk) [inline cmake/modules/FindMayaUsd.cmake]: This should be done in `cmake\test.cmake`

File: cmake/modules/FindMayaUsd.cmake [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+else()
+    if(DEFINED MAYAUSD_LOCATION)
+        message(WARNING "Could not find MayaUsd library and include directories in the folder provided : ${MAYAUSD_LOCATION}")
+    endif()
```
Comment (seando-adsk) [inline cmake/modules/FindMayaUsd.cmake]: Do you need this message? Because you didn't use `QUIET` in the `find_package(MayaUsd)` line, cmake should output a message if the package is not found.

File: lib/mayaHydra/hydraExtensions/CMakeLists.txt [hunk 26]
```diff
@@ -30,52 +30,30 @@ set(HEADERS
...
-        PRIVATE 
-            MAYAHYDRALIB_EXPORT
-            $<$<BOOL:${IS_MACOSX}>:OSMac_>
-            # Copy-pasted from lib/mayaUsd/CMakeLists.txt
-    )
-endif()
+target_compile_definitions(${TARGET_NAME} 
+    PRIVATE 
+        MAYAHYDRALIB_EXPORT
+        $<$<BOOL:${MAYAUSDAPI_LIBRARY}>:MAYAHYDRALIB_MAYAUSDAPI_ENABLED>
```
Comment (seando-adsk) [inline lib/mayaHydra/hydraExtensions/CMakeLists.txt]:
```suggestion
        $<$<BOOL:${MayaUsd_FOUND}>:MAYAHYDRALIB_MAYAUSDAPI_ENABLED>
```

File: lib/mayaHydra/hydraExtensions/CMakeLists.txt [hunk 27]
```diff
@@ -30,52 +30,30 @@ set(HEADERS
...
             PUBLIC
                 ${MAYA_INCLUDE_DIRS}
                 ${PYTHON_INCLUDE_DIR}
                 ${PXR_INCLUDE_DIRS}
                 ${BOOST_INCLUDE_DIR}
                 ${CMAKE_BINARY_DIR}/include
-        ${MAYAUSD_LOCATION}/include
             PRIVATE
                 $<$<BOOL:${UFE_FOUND}>:${UFE_INCLUDE_DIR}>
+                $<$<BOOL:${MAYAUSDAPI_LIBRARY}>:${MAYAUSD_INCLUDE_DIR}>
```
Comment (seando-adsk) [inline lib/mayaHydra/hydraExtensions/CMakeLists.txt]:
```suggestion
                $<$<BOOL:${MayaUsd_FOUND}>:${MAYAUSD_INCLUDE_DIR}>
```

File: lib/mayaHydra/hydraExtensions/CMakeLists.txt [hunk 28]
```diff
@@ -113,77 +91,40 @@ if(DEFINED MAYAHYDRA_VERSION)
...
+        usdSkel
+        usdUtils
+        ufeExtensions
+        flowViewport
+        ${MAYA_LIBRARIES}
+    PRIVATE
+        $<$<BOOL:${USD_001905_BUILD}>:hio>
+        ${PYTHON_LIBRARIES} 
+        $<$<BOOL:${UFE_FOUND}>:${UFE_LIBRARY}>
+        $<$<BOOL:${MAYAUSDAPI_LIBRARY}>:${MAYAUSDAPI_LIBRARY}>
```
Comment (seando-adsk) [inline lib/mayaHydra/hydraExtensions/CMakeLists.txt]:
```suggestion
        $<$<BOOL:${MayaUsd_FOUND}>:${MAYAUSDAPI_LIBRARY}>
```

File: test/lib/mayaUsd/render/mayaToHydra/CMakeLists.txt [hunk 29]
```diff
@@ -27,6 +27,11 @@ set(TEST_SCRIPT_FILES
     cpp/testFlowViewportAPIFilterPrims.py
 )
 
+#Add this test only if the MAYAUSDAPI_LIBRARY has been found during compile time.
+if(MAYAUSDAPI_LIBRARY)
```
Comment (seando-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/CMakeLists.txt]: The reason is because `MAYAUSDAPI_LIBRARY` is a required variable so it must be set when `MayaUsd_FOUND` is true.
```suggestion
if(MayaUsd_FOUND)
```

**Review by seando-adsk — CHANGES_REQUESTED** [review]
Looking good. Just a few minor changes.

File: cmake/test.cmake [hunk 30]
```diff
@@ -1,5 +1,15 @@
 set(MAYA_USD_DIR ${CMAKE_CURRENT_SOURCE_DIR})
 
+if(MayaUsd_FOUND)
```
Comment (seando-adsk) [inline cmake/test.cmake:3]: So here this should be either `MAYAUSD_MOD_PATH` or if you must have MAYAUSD_MOD_PATH then add it to the `REQUIRED_VARS` in FindMayaUsd.cmake.

File: cmake/test.cmake [hunk 31]
```diff
@@ -1,5 +1,15 @@
...
+    #Add MAYAUSD_MOD_PATH (the path where maya USD .mod file is) to the MAYA_MODULE_PATH
+    # Get the current value of the environment variable
+    set(CURRENT_MAYA_MODULE_PATH $ENV{MAYA_MODULE_PATH})
+    # Append the new path to the current value
+    set(MAYA_MODULE_PATH "${CURRENT_MAYA_MODULE_PATH}:${MAYAUSD_MOD_PATH}")
```
Comment (seando-adsk) [inline cmake/test.cmake]: Does Windows accept `:` in this env var? I think it would be `;` for Windows and `:` for Mac/Linux.

File: cmake/modules/FindMayaUsd.cmake (hunk 25 above)
Comment (seando-adsk) [reply cmake/modules/FindMayaUsd.cmake]: Can you test this by comment this out and then using an invalid path for `MAYAUSD_LOCATION` - I think cmake should print a warning for you.

**Review by seando-adsk — CHANGES_REQUESTED** [review]
Just one minor change left and one question to answer.

File: cmake/test.cmake [hunk 32]
```diff
@@ -5,9 +5,16 @@ if(MayaUsd_FOUND)
     # Get the current value of the environment variable
     set(CURRENT_MAYA_MODULE_PATH $ENV{MAYA_MODULE_PATH})
     # Append the new path to the current value
-    set(MAYA_MODULE_PATH "${CURRENT_MAYA_MODULE_PATH}:${MAYAUSD_MOD_PATH}")
+    if(IS_MACOSX OR IS_LINUX) 
+        #Linux and OSX
+        set(MAYA_MODULE_PATH "${CURRENT_MAYA_MODULE_PATH}:${MAYAUSD_MOD_PATH}")
+    else()
+        #Windows
+        set(MAYA_MODULE_PATH "${CURRENT_MAYA_MODULE_PATH};${MAYAUSD_MOD_PATH}")
```
Comment (seando-adsk) [inline cmake/test.cmake]: Sorry I should have remembered this earlier. The ; needs to be escaped otherwise cmake will interpret it to mean the string is a list. This is documented on line 78 in this file and used that way on line 343.

**Review by seando-adsk — CHANGES_REQUESTED** [review]
One small change needed in the test.cmake

**Review by seando-adsk — DISMISSED** [review]

File: lib/mayaHydra/mayaPlugin/renderOverride.cpp [hunk 33]
```diff
@@ -808,7 +810,7 @@ void MtohRenderOverride::_InitHydraResources(const MHWRender::MDrawContext& draw
     _initializationSucceeded = true;
 }
 
-void MtohRenderOverride::ClearHydraResources()
+void MtohRenderOverride::ClearHydraResources(bool fullReset)
```
Comment (ppt-adsk) [inline lib/mayaHydra/mayaPlugin/renderOverride.cpp:816]: Would be very valuable to have a code comment explaining how to use this new argument.

File: lib/mayaHydra/mayaPlugin/renderOverride.cpp [hunk 34]
```diff
@@ -894,7 +901,8 @@ void MtohRenderOverride::_RemovePanel(MString panelName)
     }
 
     if (_renderPanelCallbacks.empty()) {
-        ClearHydraResources();
+        static const bool fullReset = false;
```
Comment (ppt-adsk) [inline lib/mayaHydra/mayaPlugin/renderOverride.cpp]: With modern C++ this should be
```suggestion
        constexpr bool fullReset = false;
```

File: lib/mayaHydra/mayaPlugin/renderOverride.cpp [hunk 35]
```diff
@@ -1265,7 +1273,8 @@ void MtohRenderOverride::_ClearHydraCallback(void* data)
     if (!TF_VERIFY(instance)) {
         return;
     }
-    instance->ClearHydraResources();
+    static const bool fullReset = true;
```
Comment (ppt-adsk) [inline lib/mayaHydra/mayaPlugin/renderOverride.cpp]:
```suggestion
    constexpr bool fullReset = true;
```

**Review by ppt-adsk — DISMISSED** [review]
Some minor comments which you can choose to do or not.

**Review by ppt-adsk — APPROVED** [review]
